maint/local_python/binding_c.py

```diff
@@ -584,6 +675,50 @@ def dump_copy_right():
     G.out.append(" */")
     G.out.append("")
 
+def dump_profiling_qmpi(func, mapping):
+    parameters = ""
+    for p in func['c_parameters']:
+        parameters = parameters + ", " + p['name']
+
+    func_name = get_function_name(func, mapping)
+    s = get_declare_function(func, mapping)
+
+    G.out.append("#ifdef ENABLE_QMPI")
+    dump_line_with_break(s)
+    G.out.append("{")
+    G.out.append("    QMPI_Context context;")
+    G.out.append("    Q%s_t *fn_ptr;" % (func_name))
+    G.out.append("")
+    G.out.append("    context.storage_stack = NULL;")
+    G.out.append("")
+    if func_name == "MPI_Pcontrol":
+        G.out.append("    va_list varargs;")
+        G.out.append("    va_start(varargs, level);")
+        G.out.append("")
+    if "initcheck" in func['skip']:
+        G.out.append("    /* Initialize the tools that have been registered */")
+        G.out.append("    for (int i = 1; i < MPIR_QMPI_num_tools + 1; i++) {")
+        G.out.append("        MPIR_QMPI_tool_init_callbacks[i](i);")
+        G.out.append("    }")
+        G.out.append("    int mpi_errno = MPI_SUCCESS;")
+        G.out.append("    mpi_errno = MPII_qmpi_setup();")
+        G.out.append("    if (mpi_errno != MPI_SUCCESS) {")
+        G.out.append("        return mpi_errno;")
+        G.out.append("    }")
+        G.out.append("")
+        G.out.append("    MPII_qmpi_stash_first_tools();")
```

Comment:
I don't understand why you need this either. Shouldn't `MPII_qmpi_setup()` done this already?

Reply:
You have to wait until the tools have been registered to do this. This is supposed to optimize the lookup of the first function/tool ID when they're needed later. The problem comes when you have one tool that registers one subset of functions and another tool that registers another subset of functions. If you only cached the calls from the first tool, you'd miss the functions that didn't get registered for that one.

You could potentially go over the entire list of MPI functions each time, but it seems silly to do that every time a tool is registered when you can just do it once at the end of the registration period.